src/snowcli/cli/nativeapp/version/version_processor.py

```diff
@@ -0,0 +1,312 @@
+import logging
+from pathlib import Path
+from sys import stdin, stdout
+from textwrap import dedent
+from typing import Dict, List, Optional
+
+import typer
+from click import ClickException
+from git import Repo
+from git.exc import InvalidGitRepositoryError
+from rich import print
+from snowcli.cli.nativeapp.artifacts import find_version_info_in_manifest_file
+from snowcli.cli.nativeapp.constants import VERSION_COL
+from snowcli.cli.nativeapp.exceptions import ApplicationPackageDoesNotExistError
+from snowcli.cli.nativeapp.manager import (
+    NativeAppCommandProcessor,
+    NativeAppManager,
+    ensure_correct_owner,
+)
+from snowcli.cli.nativeapp.policy import PolicyBase, TaskContext
+from snowcli.cli.nativeapp.run_processor import NativeAppRunProcessor
+from snowcli.cli.nativeapp.utils import (
+    find_all_rows,
+    find_first_row,
+    is_user_in_interactive_mode,
+)
+from snowcli.cli.project.util import unquote_identifier
+from snowcli.exception import SnowflakeSQLExecutionError
+from snowflake.connector import ProgrammingError
+from snowflake.connector.cursor import DictCursor
+
+# Custom lambdas to plug in to UserConfirmationPolicy
+is_force_lambda = lambda force: force
+is_interactive_mode_lambda = lambda: stdin.isatty() and stdout.isatty()
+confirm_with_user_lambda = lambda prompt: typer.confirm(prompt)
+
+log = logging.getLogger(__name__)
+
+
+def check_index_changes_in_git_repo(
+    project_root: Path,
+    policy: PolicyBase,
+) -> None:
+    """
+    Checks if the project root, i.e. the native apps project is a git repository. If it is a git repository,
+    it also checks if there any local changes to the directory that may not be on the app package stage.
+    """
+    try:
+        repo = Repo(project_root)
+
+        # Check if the repo has any changes, including untracked files
+        if repo.is_dirty(untracked_files=True):
+            print("Changes detected in your git repository!")
+            # show differences between current files and last commit
+            print(repo.git.diff(repo.head.commit.tree))
+
+            ctx = TaskContext(
+                user_prompt="You have local changes in this repository that are not part of a previous commit. Do you still want to continue?",
+                is_interactive_mode=is_user_in_interactive_mode(),
+            )
+
+            if not policy.should_proceed(ctx):
+                if policy.exit_code:
+                    print(
+                        "Cannot create a new version non-interactively without --force"
+                    )
+                else:
+                    print("Not creating a new version.")
+                raise typer.Exit(policy.exit_code)
```

Comment:
Not fully understanding this. We're exiting either. What's the difference/why do we have different messages when exit_code is set? It's a bit odd to me that the policy can make a "and you must now exit" recommendation anyway.

Reply:
Exit codes also based on the functional specs:
1. User choosing to abort the rest of execution is exit code zero, i.e. success because that's what the user wants - `"Not creating a new version."`
2. If the execution aborts because it needed user input but was not able to proceed, that is exit code one, i.e. failure because user did not get to achieve the end state they desired - `"Cannot create a new version non-interactively without --force"`

I don't necessarily see it as the policy making a "you must now exit" recommendation. It informs the exit (on the type of exit code to use if exiting), but the invoking method is ultimately deciding when to exit. Example: in case of allow_always policy, exit code is 0 by default. But it does not tell the invoking method to exit, and the method continues execution.

Reply:
Aborting usually means exit code wouldn't be zero, even if the user chose to abort. It's still not a success. Most often the exit code is used to determine whether the next command can proceed. 

Also, the error message seems to imply that creating a version always requires --force to be provided in a CI/CD context. Is that really true?